ruby-app/lib/gamertag.rb

```diff
@@ -1,32 +1,76 @@
-module Gamertag
-  # Snowflakes, and I'm too lazy to improve these.
-  TYPE_1 = /^((ps[4|n]|xb1|xboxone)(([:-]\W*)|\W*))+/i
-  TYPE_2 = /\W*(warlock|titan|teacher)\W*ps[4|n]\W*\s?/i
+require 'strscan'
 
+module Gamertag
   def self.parse(title)
     case title
     when /slacker/i
       'kurzinator'
-    when /psn: palefacex . i'm on gmt/i
-      'PalefaceX'
-    when /software engineer at everlane; psn: esherido/i
-      'Esherido'
     when /NeuronBasher, XB1: Neuron Basher. I also run a little startup called Operable./i
       'NeuronBasher'
     when /XB1: changelog PSN: superdealloc \(CET time\)/i
       'superdealloc'
     when /Software Engineer at Heroku; PSN\/XB1: daneharrigan/i
       'daneharrigan'
-    when /PS4: RebelSenator. Titan main./i
-      'RebelSenator'
-    when /Dean of Students, Xbox One: KaiserHughes/i
-      'KaiserHughes'
-    when TYPE_1
-      title.gsub(TYPE_1, '').chomp(' ')
-    when TYPE_2
-      title.gsub(TYPE_2, '').chomp(' ')
+    when /OwlBoy - Warlock \| PS4/i
+      'OwlBoy'
     else
-      nil
+      parsers = [PlayStationParser, XboxParser]
+      parsers.each do |parser_class|
+        parser = parser_class.new(title)
+        result = parser.extract_gamertag()
+
+        return result unless result.nil?
+      end
+
+      return nil
+    end
+  end
+
+  class GamertagParser
+    attr_reader :text
+
+    def initialize(text)
+      @text = text
+    end
+
+    def extract_gamertag
+      return if self.text.nil?
+
+      scanner = StringScanner.new(self.text)
+      skipped_characters = scanner.skip_until(beginning_regex())
+      return if skipped_characters.nil?
+
+      return scanner.scan_until(ending_regex()).strip().chomp('.')
+    end
+
+    protected
+
+    def beginning_regex
```

Comment:
I was wondering this too. This is one case where an ivar or a constant may be better - I also wish Ruby had protocols or contracts or whatever because that would help here too. :)

I went with a function in this case so that I could throw an exception if the base gamertag class wasn't subclasses correctly, but this could certainly be changed.